server/rating_service/game_rater.py

```diff
@@ -15,32 +17,99 @@ class GameRatingError(Exception):
 
 @with_logger
 class GameRater:
-    @classmethod
-    def compute_rating(cls, rating_data: GameRatingData) -> Dict[PlayerID, Rating]:
-        rating_groups = [team.ratings for team in rating_data]
-        team_outcomes = [team.outcome for team in rating_data]
-        ranks = cls._ranks_from_team_outcomes(team_outcomes)
+    def __init__(self, summary: GameRatingSummary):
+        self.summary = summary
+        self.outcome_map = {
+            player_id: team.outcome
+            for team in summary.teams
+            for player_id in team.player_ids
+        }
+        self.player_ids = list(self.outcome_map.keys())
+        self.team_outcomes = [team.outcome for team in summary.teams]
+        self.ranks = _ranks_from_team_outcomes(self.team_outcomes)
+
+    def compute_rating(
+        self,
+        ratings: Dict[PlayerID, Rating]
+    ) -> Dict[PlayerID, Rating]:
+        rating_groups = [
+            {
+                player_id: trueskill.Rating(*ratings[player_id])
+                for player_id in team.player_ids
+            }
+            for team in self.summary.teams
+        ]
 
-        cls._logger.debug("Rating groups: %s", rating_groups)
-        cls._logger.debug("Ranks: %s", ranks)
+        self._logger.debug("Rating groups: %s", rating_groups)
+        self._logger.debug("Ranks: %s", self.ranks)
 
-        new_rating_groups = trueskill.rate(rating_groups, ranks)
+        new_rating_groups = trueskill.rate(rating_groups, self.ranks)
 
         player_rating_map = {
-            player_id: new_rating
+            player_id: Rating(*new_rating)
             for team in new_rating_groups
             for player_id, new_rating in team.items()
         }
 
         return player_rating_map
 
-    @staticmethod
-    def _ranks_from_team_outcomes(outcomes: List[GameOutcome]) -> List[int]:
-        if outcomes == [GameOutcome.DRAW, GameOutcome.DRAW]:
-            return [0, 0]
-        elif outcomes == [GameOutcome.VICTORY, GameOutcome.DEFEAT]:
-            return [0, 1]
-        elif outcomes == [GameOutcome.DEFEAT, GameOutcome.VICTORY]:
-            return [1, 0]
-        else:
-            raise GameRatingError(f"Inconsistent outcomes {outcomes}")
+    def get_outcome_map(self) -> Dict[PlayerID, GameOutcome]:
+        return self.outcome_map
+
+
+@with_logger
+class AdjustmentGameRater(GameRater):
+    """GameRater for performing adjustments using another GameRater"""
+
+    def __init__(self, rater: GameRater, base_ratings: Dict[PlayerID, Rating]):
+        self.rater = rater
+        self.base_ratings = base_ratings
+
+    def compute_rating(
+        self,
+        ratings: Dict[PlayerID, Rating]
+    ) -> Dict[PlayerID, Rating]:
+        """
+        Perform rating adjustment based on another set of ratings. For
+        each player, this will rate the game with trueskill as if their
+        rating for the base rating type were equal to the rating we are
```

Comment:
```suggestion
        each player, this will rate the game with trueskill as if they
        played this game with the rating we are adjusting instead of the
        base rating.
```
This makes it clearer what is happening imo

Reply:
Haha, yea I refactored it to be more generic and then failed epically to update the comments